### ISSUE TITLE
pthread-main.js: FS is not defined

### ISSUE THREAD

**Issue by loganmc10** [kripken/emscripten#4578]
I am trying to use pthreads, but I get an error saying that FS is not defined in pthread-main.js.

Is there any way to work around this? I assume this line:

https://github.com/kripken/emscripten/blob/master/src/pthread-main.js#L50

is the issue

**Comment by juj** [comment]
Fixed in https://github.com/kripken/emscripten/commit/8c2e31ad6a233f521938209bddc91d26ca94cf22. That should be enough to avoid referencing `FS` anywhere later either, since that single instance is hardcoded due to stream objects not sharing state between main thread and workers.

**Comment by loganmc10** [comment]
I also get an error that says "_usleep is not a function" when trying to use pthreads, have you ever heard of others running into that issue as well?

**Comment by juj** [comment]
I do recall that coming up in the past. https://linux.die.net/man/3/usleep mentions that

    4.3BSD, POSIX.1-2001. POSIX.1-2001 declares this function obsolete;
    use nanosleep(2) instead. POSIX.1-2008 removes the specification of usleep().

Nanosleep should exist in Emscripten. We could consider emulating `usleep`, though quickest is probably to convert to `nanosleep`.

**Comment by loganmc10** [comment]
ok ill try that, thanks

**Comment by juj** [comment]
Hmm actually, `usleep` is implemented in https://github.com/kripken/emscripten/blob/incoming/system/lib/pthread/library_pthread.c#L250 for pthreads code, and it is exported in https://github.com/kripken/emscripten/blob/incoming/system/lib/pthreads.symbols#L309, so not quite sure why that is not found.

**Comment by loganmc10** [comment]
Ya the actual stacktrace looks like this:
```
Uncaught TypeError: _usleep is not a function
_nanosleep @ VM233:10843
_emscripten_mainloop @ VM233:118080
dynCall_v @ VM233:612663
dynCall @ VM233:336(anonymous function) @ VM233:2703
runIter @ VM233:2765Browser_mainLoop_runner @ VM233:2699
```

The .js file has this:
```
var _usleep=undefined;function _nanosleep(rqtp, rmtp) {
      // int nanosleep(const struct timespec  *rqtp, struct timespec *rmtp);
      var seconds = HEAP32[((rqtp)>>2)];
      var nanoseconds = HEAP32[(((rqtp)+(4))>>2)];
      if (rmtp !== 0) {
        HEAP32[((rmtp)>>2)]=0;
        HEAP32[(((rmtp)+(4))>>2)]=0;
      }
      return _usleep((seconds * 1e6) + (nanoseconds / 1000));
    }
```

And this is emscripten_mainloop()
```
function _emscripten_mainloop() {
 var $0 = 0, $1 = 0, $10 = 0, $11 = 0, $12 = 0, $2 = 0, $3 = 0, $4 = 0, $5 = 0, $6 = 0, $7 = 0, $8 = 0, $9 = 0, $or$cond = 0, $sleep_ms = 0, $tv$i = 0, label = 0, sp = 0;
 sp = STACKTOP;
 STACKTOP = STACKTOP + 16|0; if ((STACKTOP|0) >= (STACK_MAX|0)) abort();
 $tv$i = sp;
 $sleep_ms = sp + 8|0;
 HEAP32[$sleep_ms>>2] = 0;
 $0 = (_runloop_iterate($sleep_ms)|0);
 $1 = ($0|0)==(1);
 $2 = HEAP32[$sleep_ms>>2]|0;
 $3 = ($2|0)!=(0);
 $or$cond = $1 & $3;
 if ($or$cond) {
  $4 = $tv$i;
  $5 = $4;
  HEAP32[$5>>2] = 0;
  $6 = (($4) + 4)|0;
  $7 = $6;
  HEAP32[$7>>2] = 0;
  $8 = (($2>>>0) / 1000)&-1;
  HEAP32[$tv$i>>2] = $8;
  $9 = (($2>>>0) % 1000)&-1;
  $10 = ($9*1000000)|0;
  $11 = ((($tv$i)) + 4|0);
  HEAP32[$11>>2] = $10;
  (_nanosleep(($tv$i|0),(0|0))|0);
 }
 (_task_queue_ctl(6,0)|0);
 $12 = ($0|0)==(-1);
 if ($12) {
  _main_exit(0);
  _exit(0);
  // unreachable;
 } else {
  STACKTOP = sp;return;
 }
}
```

**Comment by juj** [comment]
Oh, thanks, now I understand what was going on. Fix in pull #4579.

**Comment by loganmc10** [comment]
great, thanks!
